pyface/data_view/abstract_value_type.py

```diff
@@ -0,0 +1,180 @@
+# (C) Copyright 2005-2020 Enthought, Inc., Austin, TX
+# All rights reserved.
+#
+# This software is provided without warranty under the terms of the BSD
+# license included in LICENSE.txt and may be redistributed only under
+# the conditions described in the aforementioned license. The license
+# is also available online at http://www.enthought.com/licenses/BSD.txt
+#
+# Thanks for using Enthought open source!
+
+""" Provides an AbstractValueType ABC for Pyface data models.
+
+This module provides an ABC for data view value types, which are responsible
+for adapting raw data values as used by the data model's ``get_value`` and
+``set_value`` methods to the data channels that the data view expects, such
+as text, color, icons, etc.
+
+It is up to the data view to take this standardized data and determine what
+and how to actually display it.
+"""
+
+from traits.api import ABCHasStrictTraits, Event, Str, observe
+
+
+class AbstractValueType(ABCHasStrictTraits):
+    """ A value type converts raw data into data channels.
+
+    The data channels are editor value, text, color, image, and description.
+    The data channels are used by other parts of the code to produce the actual
+    display.
+
+    Subclasses should mark traits that potentially affect the display of values
+    with ``update=True`` metdadata, or alternativelym fire the ``updated``
+    event when the state of the value type changes.
+    """
+
+    #: Fired when a change occurs that requires updating values.
+    updated = Event
+
+    def can_edit(self, model, row, column):
+        """ Return whether or not the value can be edited.
+
+        The default implementation is that cells that can be set are
+        editable.
+
+        Parameters
+        ----------
+        model : AbstractDataModel
+            The data model holding the data.
+        row : sequence of int
+            The row in the data model being queried.
+        column : sequence of int
+            The column in the data model being queried.
+
+        Returns
+        -------
+        can_edit : bool
+            Whether or not the value is editable.
+        """
+        return model.can_set_value(row, column)
+
+    def get_editable(self, model, row, column):
+        """ Return a value suitable for editing.
+
+        The default implementation is to return the underlying data value
+        directly from the data model.
+
+        Parameters
+        ----------
+        model : AbstractDataModel
+            The data model holding the data.
+        row : sequence of int
+            The row in the data model being queried.
+        column : sequence of int
+            The column in the data model being queried.
+
+        Returns
+        -------
+        value : any
+            The value to edit.
+        """
+        return model.get_value(row, column)
```

Comment:
Yes, that is the intent, and the "user" in this case is going to be internal Pyface toolkit code.

I did try some alternatives: I tried raising an exception initially, but it is too expensive for something being called repeatedly; and I considered `None` as a null value (which _would_ work for text), but experience has taught me that we will inevitably end up with a situation where `None` is somehow a valid `editable` value.

So I settled on a "look before you leap" pattern with a "can I get it" followed by a "please get it".  There are probably more elegant ways of doing it, but it matches the python hasattr/getattr/setattr pattern.

Reply:
Okay, that makes sense, but I still think that this should be documented somewhere in the code so that the information doesn't get lost. Maybe the docstring of `AbstractValueType` would be a good place?